### DISCUSSION TITLE
JSON extension in Python

### DISCUSSION THREAD

**Discussion by rgieseke** [duckdb/duckdb#3267] [Q&A] [answered]
In the `setup.py` the JSON extension is [listed](https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/setup.py#L13) but I can't get the examples to run.

```py
import duckdb

duckdb.__version__

'0.3.3-dev585'

con = duckdb.connect(":memory:")

con.execute("""SELECT to_json('duck');""")

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 con.execute("""SELECT to_json('duck');""")

RuntimeError: Catalog Error: Scalar Function with name to_json does not exist!
Did you mean "to_days"?
LINE 1: SELECT to_json('duck');
               ^

con.execute("""CREATE TABLE example (j JSON);""")

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 con.execute("""CREATE TABLE example (j JSON);""")

RuntimeError: Not implemented Error: DataType JSON not supported yet...

```

Is there anything else needed to use the extension?

**Comment by Alex-Monahan** [comment]
I haven't tested this out yet, but what if you run 
```sql
LOAD 'json';
```
before your other queries?
[Extension docs](https://duckdb.org/docs/extensions/overview)

**Reply by rgieseke** [reply]
Thanks for the quick reply!

This gives

```RuntimeError                              Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 con.execute("""LOAD 'json';""")

RuntimeError: IO Error: File "/home/robert/.duckdb/extensions/b52293246/linux_amd64/json.duckdb_extension" not found
```

I did try some lines from the `fts` extension's example which worked fine.

**Reply by Alex-Monahan** [reply]
Could you provide your setup details (OS, etc. )?

**Comment by Alex-Monahan** [comment]
I am out of troubleshooting steps! 
@lnkuiper, could you have a look? Thank you!!

**Reply by rgieseke** [reply]
> Could you provide your setup details (OS, etc. )?

Sure! I'm trying a local build now, looks like there has been lots of activity in the JSON extension recently!

I'm on Ubuntu Linux, Python 3.9.7, `duckdb` installed into a virtualenv with `pip install duckdb --pre`. 
`0.3.3-dev585`
Looks like the latest Pre-Releases on PyPI only have Windows wheels. (Not sure though if anything JSON related happened in the commits in-between.)

**Reply by rgieseke** [reply]
FWIW the build seems to succeed (after I commented out the `substrait` extension in `setup.py`, but on import in a virtual env i get:

```
[...]/venv/lib/python3.9/site-packages/duckdb-0.3.3.dev1094+g4eb59236e.d20220320-py3.9-linux-x86_64.egg/duckdb.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZN6duckdb10FileSystem4GlobERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS_10FileOpenerE
```

**Reply by lnkuiper** [reply]
Can you try
```sql
INSTALL 'json';
LOAD 'json';
```

**Reply by rgieseke** [reply]
For ` 0.3.3-dev585` I get
```
----> 1 con.execute("""INSTALL 'json';""")

RuntimeError: IO Error: Failed to download extension http://extensions.duckdb.org/b52293246/linux_amd64/json.duckdb_extension.gz
```

For regular 3.2:
```
RuntimeError: IO Error: Failed to download extension http://extensions.duckdb.org/5aebf7dac/linux_amd64/json.duckdb_extension.gz
```

**Reply by hallenmaia** [reply]
```python
conn.execute("INSTALL 'JSON'")
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
RuntimeError: IO Error: Failed to download extension http://extensions.duckdb.org/5aebf7dac/osx_arm64/JSON.duckdb_extension.gz
```

Same error on Mac M1 for 0.3.2

**Comment by hallenmaia** [comment]
If I access the url through the browser, it displays this error:

<img width="775" alt="Captura de Tela 2022-03-20 às 15 10 56" src="https://user-images.githubusercontent.com/29228176/159176448-30f4ead5-6961-4df8-8977-00a766387c58.png">

**Comment by lnkuiper** [comment, marked as answer]
Thank you for the detailed error report. For 0.3.2 it is not able to install the JSON extension yet, because it was added after the 0.3.2 release. You can expect the JSON extension to appear in the next release.

For the latest master builds of duckdb installing the JSON extension SHOULD work, make sure the version you're using is a version later than PR #3214 was merged, as this fixed extension linking issues. Looking at the release dates on [PyPI](https://pypi.org/project/duckdb/#history), this is not yet possible.

If you wish to use the JSON extension already, you will have to wait until this version is uploaded to PyPI, or build from source yourself. Execute the following command in the duckdb directory after cloning the repo:
```bash
BUILD_PYTHON=1 BUILD_JSON=1 DUCKDB_INSTALL_USER=1 make
```

Building from source takes some time, so I recommend installing `ninja-build` with your package manager, and adding `GEN=ninja` to the command above to parallelise the build.

After building from source, you should be able to use the JSON extension without any `INSTALL/LOAD` commands.

Sorry for the inconvenience, the JSON extension will be more accessible soon!

**Reply by rgieseke** [reply]
Thank you!

For installing in a virtualenv the following worked for me (no user install and added substrait):
```
GEN=ninja BUILD_PYTHON=1 BUILD_JSON=1 BUILD_SUBSTRAIT_EXTENSION=1  make
```

Thanks everyone for the helpful replies (and no need to apologise, exploring DuckDB has been great fun and i realise i was really playing with very recently added features!)
